specifyweb/frontend/js_src/lib/components/FormSliders/IntegratedRecordSelector.tsx

```diff
@@ -173,48 +57,61 @@ export function IntegratedRecordSelector({
   const index = f.parseInt(rawIndex) ?? 0;
+
+  const isInteraction =
+    typeof relationship === 'object' &&
+    relationshipIsToMany(relationship) &&
+    typeof collection.related === 'object' &&
+    ['LoanPreparation', 'GiftPreparation', 'DisposalPreparation'].includes(
+      relationship.relatedTable.name
+    );
+
+  const [isDialogOpen, handleOpenDialog, handleCloseDialog] = useBooleanState();
+
   return (
     <ReadOnlyContext.Provider value={isReadOnly}>
-      {formType === 'formTable' ? (
-        <FormTableCollection
-          collection={collection}
-          dialog={dialog}
-          sortField={sortField}
-          viewName={viewName}
-          onAdd={(resources): void => {
-            collection.add(resources);
-            if (typeof handleAdd === 'function') handleAdd(resources);
-          }}
-          onClose={handleClose}
-          onDelete={
-            handleDelete === undefined
-              ? undefined
-              : (_resource, index): void => handleDelete(index, 'minusButton')
-          }
-        />
-      ) : (
-        <RecordSelectorFromCollection
-          collection={collection}
-          defaultIndex={isToOne ? 0 : index}
-          relationship={relationship}
-          onAdd={handleAdd}
-          onDelete={handleDelete}
-          onSlide={(index): void =>
-            typeof urlParameter === 'string'
-              ? setIndex(index.toString())
-              : undefined
+      <RecordSelectorFromCollection
+        collection={collection}
+        defaultIndex={isToOne ? 0 : index}
+        relationship={relationship}
+        isInteraction={isInteraction}
+        onAdd={(resource) => {
+          handleOpenDialog();
+          if (handleAdd !== undefined && isInteraction) {
+            handleAdd(resource);
           }
-          {...rest}
-        >
-          {({
-            dialogs,
-            slider,
-            resource,
-            onAdd: handleAdd,
-            onRemove: handleRemove,
-            isLoading,
-          }): JSX.Element => (
-            <>
+        }}
+        onDelete={handleDelete}
+        onSlide={(index): void =>
+          typeof urlParameter === 'string'
+            ? setIndex(index.toString())
+            : undefined
+        }
+        {...rest}
+      >
+        {({
+          dialogs,
+          slider,
+          resource,
+          onAdd: handleAdd,
+          onRemove: handleRemove,
+          isLoading,
+        }): JSX.Element => (
+          <>
+            {isInteraction &&
+            typeof collection.related === 'object' &&
+            isDialogOpen ? (
+              <InteractionDialog
+                actionTable={collection.related.specifyTable}
+                itemCollection={
+                  collection as Collection<
+                    DisposalPreparation | GiftPreparation | LoanPreparation
+                  >
+                }
+                onClose={handleCloseDialog}
```

Comment:
Regarding `you can fix it by adding a prop to the prep select dialog`:

When you display the InteractionDialog, two things could happen:
1. Either the user closes the dialog, without selecting anything - in this case only `handleCloseDialog` is called
2. OR, the user selects new props and then the dialog closes - in this case new records are added to `collection` and `handleCloseDialog` is closed.

What you need to do is when the dialog is closed, figure out what is the index of the first new added record (i.e collection had 3 records (thus last index is `2`), and user added 5 more, then the index of first added is `3`). And then you would open the record with that index in the subview

Some possible ways to do that
1. Add a callback i.e `onAdded={}` that would be called by InteractionDialog when it adds new records. the callback could be called right before new records are added (thus you could use `collection.models.length` as new index) or right after (in which case InteractionDialog will pass the new records, and you could do `collection.models.length - addedRecords.length` to get new index)
2. OR record the number of records in the collection before the interaction dialog is opened in a ref, and once the interaction dialog is closed, check if number of records changes and get the new index
3. OR, set an event listener for new items being added, and change the index when that happens - though this one would be tricky as if 4 records are added all at once, the event index would be called 4 times, thus, subview will end up displaying the last record

Reply:
new issue created :+1: